telemetry/record_test.go

```diff
@@ -27,56 +34,185 @@ var _ = Describe("Socket handler test", func() {
+	})
+
+	It("passes through a valid Location location", func() {
+		expected := &protos.LocationValue{Latitude: 37.412374, Longitude: -122.145867}
+		loc := locationDatum(protos.Field_Location, expected)
+
+		message := messages.StreamMessage{TXID: []byte("1234"), SenderID: []byte("vehicle_device.42"), MessageTopic: []byte("V"), Payload: generatePayload("cybertruck", "42", nil, loc)}
+		recordMsg, err := message.ToBytes()
+		Expect(err).NotTo(HaveOccurred())
+
+		record, err := telemetry.NewRecord(serializer, recordMsg, "1")
+		Expect(err).NotTo(HaveOccurred())
+		Expect(record).NotTo(BeNil())
+
+		data := &protos.Payload{}
+		err = proto.Unmarshal(record.Payload(), data)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(data.Data).To(HaveLen(2))
+
+		sort.Slice(data.Data, func(i, j int) bool {
+			return data.Data[i].Key < data.Data[j].Key
+		})
+
+		first := data.Data[0]
+		Expect(first.Key).To(Equal(protos.Field_Location))
+		Expect(first.Value.GetStringValue()).To(BeEmpty())
+		Expect(first.Value.GetLocationValue()).To(Equal(clone(expected)))
+
+		second := data.Data[1]
+		Expect(second.Key).To(Equal(protos.Field_VehicleName))
+	})
 })
 
-func generatePayload(vehicleName string, vin string, timestamp *timestamppb.Timestamp) []byte {
+var _ = DescribeTable("ParseLocation",
+	func(locStr string, expected *protos.LocationValue, errRegex string) {
+		loc, err := telemetry.ParseLocation(locStr)
+		if errRegex == "" {
+			Expect(err).NotTo(HaveOccurred())
+		} else {
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(MatchRegexp(errRegex))
+		}
+		Expect(loc).To(Equal(clone(expected)))
+	},
+	Entry("for a bogus string", "Abhishek is NOT a location", nil, "input does not match format"),
+	Entry("for a broken location", "(37.412374 Q, 122.145867 W)", nil, "invalid location format.*"),
+	Entry("for a valid loc, NW", "(37.412374 N, 122.145867 W)", &protos.LocationValue{Latitude: 37.412374, Longitude: -122.145867}, ""),
+	Entry("for a valid loc, NE", "(37.412374 N, 122.145867 E)", &protos.LocationValue{Latitude: 37.412374, Longitude: 122.145867}, ""),
+	Entry("for a valid loc, SW", "(37.412374 S, 122.145867 W)", &protos.LocationValue{Latitude: -37.412374, Longitude: -122.145867}, ""),
+	Entry("for a valid loc, SE", "(37.412374 S, 122.145867 E)", &protos.LocationValue{Latitude: -37.412374, Longitude: 122.145867}, ""),
+)
+
+func generatePayload(vehicleName string, vin string, timestamp *timestamppb.Timestamp, extraData ...*protos.Datum) []byte {
 	var data []*protos.Datum
-	data = append(data, &protos.Datum{
-		Key: protos.Field_VehicleName,
-		Value: &protos.Value{
-			Value: &protos.Value_StringValue{
-				StringValue: vehicleName,
-			},
-		},
-	})
+	data = append(data, stringDatum(protos.Field_VehicleName, vehicleName))
+	data = append(data, extraData...)
+
 	payload, err := proto.Marshal(&protos.Payload{
 		Vin:       vin,
 		Data:      data,
 		CreatedAt: timestamp,
 	})
-	Expect(err).To(BeNil())
```

Comment:
can we convert all existing assertions to `Expect(err).NotTo(HaveOccurred())`?

Reply:
Isn't that the case already? (Or do you mean beyond `record_test`?)

Reply:
doh, I like `BeNil()`; it's shorter

Reply:
But then the failure message is more expressive...

Reply:
@agbpatro see #54 (separate PR to make thing clear).